### PR TITLE
fix: Argo CD unnecessary enforce sequential helm manifest generation for one chart

### PR DESCRIPTION
**Pull request by alexmt**
Part of https://github.com/argoproj/argo-cd/issues/17517

PR is the first step to address https://github.com/argoproj/argo-cd/issues/17517 .  TLDR: it is safe to run `helm dep up && helm template` even if helm chart has conditional dependencies. Testing did not reveal any issues.

To be safe the first I'm proposing to make concurrent generation opt-in using `ARGOCD_HELM_ALLOW_CONCURRENCY` env variable .  In the next release we can remove env variable as well as code related to `.argocd-allow-concurrency` file.

Since `ARGOCD_HELM_ALLOW_CONCURRENCY` is not meant to stay around for too long it is documented in `high_availability.md` but not in `argocd-cmd-params-cm-yaml.md`.

### CODE REVIEW

**Review by gdsoumya — APPROVED** [review]
LGTM

**Review by 34fathombelow — APPROVED** [review]
LGTM

**Review by morey-tech — APPROVED** [review]

**Review by pasha-codefresh — APPROVED** [review]

File: docs/operator-manual/high_availability.md [hunk 1]
```diff
@@ -170,8 +170,9 @@ Argo CD repo server maintains one repository clone locally and uses it for appli
 Argo CD determines if manifest generation might change local files in the local repository clone based on the config management tool and application settings.
 If the manifest generation has no side effects then requests are processed in parallel without a performance penalty. The following are known cases that might cause slowness and their workarounds:
 
-  * **Multiple Helm based applications pointing to the same directory in one Git repository:** ensure that your Helm chart doesn't have conditional
-[dependencies](https://helm.sh/docs/chart_best_practices/dependencies/#conditions-and-tags) and create `.argocd-allow-concurrency` file in the chart directory.
+  * **Multiple Helm based applications pointing to the same directory in one Git repository:** for historical reasons Argo CD generates Helm manifests
+    sequential.  To enable parallel generation set `ARGOCD_HELM_ALLOW_CONCURRENCY=true` to `argocd-repo-server` deployment or create `.argocd-allow-concurrency` file.
+    Future versions of Argo CD will enable this by default.
```
Comment (todaywasawesome) [inline docs/operator-manual/high_availability.md:175]:
```suggestion
  * **Multiple Helm based applications pointing to the same directory in one Git repository:** for historical reasons Argo CD generates Helm manifests sequentially.  To enable parallel generation set `ARGOCD_HELM_ALLOW_CONCURRENCY=true` to `argocd-repo-server` deployment or create `.argocd-allow-concurrency` file.
    Future versions of Argo CD will enable this by default.
```

**Review by todaywasawesome — APPROVED** [review]
Tweaked wording and now looks great.

File: reposerver/repository/repository.go [hunk 2]
```diff
@@ -74,7 +71,12 @@ const (
 	ociPrefix                      = "oci://"
 )
 
-var ErrExceededMaxCombinedManifestFileSize = errors.New("exceeded max combined manifest file size")
+var (
+	ErrExceededMaxCombinedManifestFileSize = errors.New("exceeded max combined manifest file size")
+	// helmConcurrencyDefault if true then helm concurrent manifest generation is enabled
+	// TODO: remove env variable and usage of .argocd-allow-concurrency once we are sure that it is safe to enable it by default
+	helmConcurrencyDefault = env.ParseBoolFromEnv("ARGOCD_HELM_ALLOW_CONCURRENCY", false)
```
Comment (crenshaw-dev) [inline reposerver/repository/repository.go:78]: Would recommend implementing this as a repo server flag w/ env var fallback and using argocd-cmd-params-cm to set/document it.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

**Comment by crenshaw-dev** [comment]
> make concurrent generation opt-in using ARGOCD_HELM_ALLOW_CONCURRENCY env variable . In the next release we can remove env variable

Maybe this release cycle opt-in, next release cycle opt-out, and then finally just remove the code? One cycle of opt-out would provide a lot more confidence in the change I think.

**Pull request merged**